File: backend/app/services/claude.py
import asyncio
import os
import shutil
import time
from typing import Any, Dict, List, Optional

from app.core.config import SESSION_OUTPUT_LINES
from app.db import jobs_repo
from app.services.sessions import (
    append_session_message,
    format_session_prompt,
    get_session_messages,
)
from app.websocket.manager import manager


def resolve_claude_path() -> Optional[str]:
    cli_path = os.environ.get("CLAUDE_CODE_CLI_PATH")
    if cli_path and os.path.exists(cli_path):
        return cli_path
    return shutil.which("claude")


async def read_stream_lines(
    stream: asyncio.StreamReader,
    level: str,
    prefix: str,
    collector: List[str],
    limit: int,
) -> None:
    while True:
        line = await stream.readline()
        if not line:
            break
        text = line.decode(errors="replace").rstrip()
        await manager.emit_log(level, f"{prefix}{text}")
        collector.append(text)
        if len(collector) > limit:
            collector.pop(0)


async def run_claude_job(job_id: str, payload: Dict[str, Any]) -> None:
    claude_path = resolve_claude_path()
    if not claude_path:
        await jobs_repo.update_job(
            job_id,
            status="failed",
            error={"message": "claude cli not found"},
        )
        await jobs_repo.record_event(job_id, "error", "claude cli not found")
        await manager.broadcast(
            {"type": "job.status", "job_id": job_id, "status": "failed"}
        )
        return

    session_id = payload.get("session_id")
    args = payload.get("args") or []
    raw_prompt = payload.get("prompt")
    prompt = raw_prompt
    if session_id and raw_prompt:
        history = await get_session_messages(session_id)
        prompt = format_session_prompt(history, raw_prompt)
    if prompt and not args:
        args = ["--print", prompt]
    if not args:
        await jobs_repo.update_job(
            job_id,
            status="failed",
            error={"message": "no args provided for claude cli"},
        )
        await jobs_repo.record_event(job_id, "error", "claude cli args missing")
        await manager.broadcast(
            {"type": "job.status", "job_id": job_id, "status": "failed"}
        )
        return

    timeout_ms = payload.get("timeout_ms")
    timeout_sec = None if timeout_ms is None else max(timeout_ms / 1000.0, 1.0)
    stdin_payload = payload.get("stdin")

    await jobs_repo.update_job(
        job_id, status="running", progress=0.05, message="spawning"
    )
    await manager.broadcast(
        {"type": "job.status", "job_id": job_id, "status": "running"}
    )

    stdout_lines: List[str] = []
    stderr_lines: List[str] = []
    started = time.time()

    process = await asyncio.create_subprocess_exec(
        claude_path,
        *args,
        stdin=asyncio.subprocess.PIPE,
        stdout=asyncio.subprocess.PIPE,
        stderr=asyncio.subprocess.PIPE,
        env=os.environ.copy(),
    )

    if stdin_payload:
        process.stdin.write(stdin_payload.encode())
        await process.stdin.drain()
    if process.stdin:
        process.stdin.close()

    stdout_task = asyncio.create_task(
        read_stream_lines(
            process.stdout, "info", "claude: ", stdout_lines, SESSION_OUTPUT_LINES
        )
    )
    stderr_task = asyncio.create_task(
        read_stream_lines(
            process.stderr, "error", "claude: ", stderr_lines, SESSION_OUTPUT_LINES
        )
    )

    try:
        if timeout_sec:
            await asyncio.wait_for(process.wait(), timeout=timeout_sec)
        else:
            await process.wait()
    except asyncio.TimeoutError:
        process.kill()
        await jobs_repo.update_job(
            job_id,
            status="failed",
            error={"message": "claude cli timeout"},
        )
        await jobs_repo.record_event(job_id, "error", "claude cli timeout")
        await manager.broadcast(
            {"type": "job.status", "job_id": job_id, "status": "failed"}
        )
        return
    finally:
        await asyncio.gather(stdout_task, stderr_task, return_exceptions=True)

    duration_ms = int((time.time() - started) * 1000)
    exit_code = process.returncode
    stdout_tail = stdout_lines[-50:] if stdout_lines else []
    stderr_tail = stderr_lines[-50:] if stderr_lines else []
    if exit_code == 0:
        if session_id and raw_prompt:
            assistant_output = "\n".join(stdout_lines).strip()
            await append_session_message(session_id, "user", raw_prompt)
            await append_session_message(session_id, "assistant", assistant_output)
        await jobs_repo.update_job(
            job_id,
            status="succeeded",
            progress=1.0,
            result={
                "exit_code": exit_code,
                "duration_ms": duration_ms,
                "stdout_tail": stdout_tail,
                "stderr_tail": stderr_tail,
            },
        )
        await jobs_repo.record_event(job_id, "info", "claude cli finished")
        await manager.broadcast(
            {"type": "job.status", "job_id": job_id, "status": "succeeded"}
        )
    else:
        await jobs_repo.update_job(
            job_id,
            status="failed",
            error={
                "message": "claude cli failed",
                "exit_code": exit_code,
                "stdout_tail": stdout_tail,
                "stderr_tail": stderr_tail,
            },
        )
        await jobs_repo.record_event(job_id, "error", "claude cli failed")
        await manager.broadcast(
            {"type": "job.status", "job_id": job_id, "status": "failed"}
        )

File: backend-bun/src/services/claude.ts
import { $ } from "bun";
import * as jobsRepo from "../db/jobs.ts";
import { config } from "../lib/config.ts";
import { manager } from "../websocket/manager.ts";
import {
  getSessionMessages,
  appendSessionMessage,
  formatSessionPrompt,
} from "./sessions.ts";

function resolveClaudePath(): string | null {
  const cliPath = Bun.env.CLAUDE_CODE_CLI_PATH;
  if (cliPath && Bun.file(cliPath).size > 0) {
    return cliPath;
  }
  // Check if claude is in PATH
  try {
    const result = Bun.spawnSync(["which", "claude"]);
    if (result.exitCode === 0) {
      return result.stdout.toString().trim();
    }
  } catch {
    // Ignore errors
  }
  return null;
}

export async function runClaudeJob(
  jobId: string,
  payload: Record<string, unknown>
): Promise<void> {
  const claudePath = resolveClaudePath();
  if (!claudePath) {
    jobsRepo.updateJob(jobId, {
      status: "failed",
      error: { message: "claude cli not found" },
    });
    jobsRepo.recordEvent(jobId, "error", "claude cli not found");
    manager.emitJobStatus(jobId, "failed");
    return;
  }

  const sessionId = payload.session_id as string | undefined;
  let args = (payload.args as string[]) || [];
  const rawPrompt = payload.prompt as string | undefined;
  let prompt = rawPrompt;

  if (sessionId && rawPrompt) {
    const history = getSessionMessages(sessionId);
    prompt = formatSessionPrompt(history, rawPrompt);
  }

  if (prompt && args.length === 0) {
    args = ["--print", prompt];
  }

  if (args.length === 0) {
    jobsRepo.updateJob(jobId, {
      status: "failed",
      error: { message: "no args provided for claude cli" },
    });
    jobsRepo.recordEvent(jobId, "error", "claude cli args missing");
    manager.emitJobStatus(jobId, "failed");
    return;
  }

  const timeoutMs = payload.timeout_ms as number | undefined;
  const timeoutSec = timeoutMs ? Math.max(timeoutMs / 1000, 1) : undefined;
  const stdinPayload = payload.stdin as string | undefined;

  jobsRepo.updateJob(jobId, { status: "running", progress: 0.05, message: "spawning" });
  manager.emitJobStatus(jobId, "running");

  const stdoutLines: string[] = [];
  const stderrLines: string[] = [];
  const started = Date.now();

  try {
    const proc = Bun.spawn([claudePath, ...args], {
      stdin: stdinPayload ? new TextEncoder().encode(stdinPayload) : undefined,
      stdout: "pipe",
      stderr: "pipe",
      env: { ...process.env },
    });

    // Create timeout promise if needed
    let timeoutId: Timer | undefined;
    if (timeoutSec) {
      timeoutId = setTimeout(() => {
        proc.kill();
      }, timeoutSec * 1000);
    }

    // Wait for process to complete
    const exitCode = await proc.exited;
    if (timeoutId) clearTimeout(timeoutId);

    // Read output
    const stdoutText = await new Response(proc.stdout).text();
    const stderrText = await new Response(proc.stderr).text();

    for (const line of stdoutText.split("\n")) {
      const trimmed = line.trim();
      if (trimmed) {
        manager.emitLog("info", `claude: ${trimmed}`);
        stdoutLines.push(trimmed);
        if (stdoutLines.length > config.SESSION_OUTPUT_LINES) {
          stdoutLines.shift();
        }
      }
    }

    for (const line of stderrText.split("\n")) {
      const trimmed = line.trim();
      if (trimmed) {
        manager.emitLog("error", `claude: ${trimmed}`);
        stderrLines.push(trimmed);
        if (stderrLines.length > config.SESSION_OUTPUT_LINES) {
          stderrLines.shift();
        }
      }
    }

    const durationMs = Date.now() - started;
    const stdoutTail = stdoutLines.slice(-50);
    const stderrTail = stderrLines.slice(-50);

    if (exitCode === 0) {
      if (sessionId && rawPrompt) {
        const assistantOutput = stdoutLines.join("\n").trim();
        appendSessionMessage(sessionId, "user", rawPrompt);
        appendSessionMessage(sessionId, "assistant", assistantOutput);
      }

      jobsRepo.updateJob(jobId, {
        status: "succeeded",
        progress: 1,
        result: {
          exit_code: exitCode,
          duration_ms: durationMs,
          stdout_tail: stdoutTail,
          stderr_tail: stderrTail,
        },
      });
      jobsRepo.recordEvent(jobId, "info", "claude cli finished");
      manager.emitJobStatus(jobId, "succeeded");
    } else {
      jobsRepo.updateJob(jobId, {
        status: "failed",
        error: {
          message: "claude cli failed",
          exit_code: exitCode,
          stdout_tail: stdoutTail,
          stderr_tail: stderrTail,
        },
      });
      jobsRepo.recordEvent(jobId, "error", "claude cli failed");
      manager.emitJobStatus(jobId, "failed");
    }
  } catch (error) {
    const errorMessage = error instanceof Error ? error.message : String(error);

    if (errorMessage === "timeout") {
      jobsRepo.updateJob(jobId, {
        status: "failed",
        error: { message: "claude cli timeout" },
      });
      jobsRepo.recordEvent(jobId, "error", "claude cli timeout");
    } else {
      jobsRepo.updateJob(jobId, {
        status: "failed",
        error: { message: errorMessage },
      });
      jobsRepo.recordEvent(jobId, "error", `claude cli error: ${errorMessage}`);
    }
    manager.emitJobStatus(jobId, "failed");
  }
}
